[GeoMechanicsApplication] Improved robustness of nodal hydraulic head tests

avdg81 opened the pull request:
**📝 Description**
The nodal hydraulic head tests are no longer implemented using the C++ unit test framework, but using the Python unit test framework, since these are integration tests. Previously, these tests were rather fragile, since they compared GiD output files, character by character. If the file contents would differ by at least one character, the test would fail.

Now we read the produced results from the GiD output files and compare the obtained hydraulic head values against the analytical solution, taking into account a tolerance value. This should make the test much more robust when the obtained results change slightly, e.g. when the precision of the produced results changes.

To run the flow analysis from a Python test script using the C++ workflow, a Python runner script has been added to the custom Python scripts directory.

avdg81 committed:
Added a Python script for running GeoFlow calculations

This script uses the C++ workflow.

avdg81 committed:
Added a Python test script for checking the nodal hydraulic heads

avdg81 committed:
Extracted a function that runs and checks the flow analysis

avdg81 committed:
Added three more test cases

avdg81 committed:
Removed now redundant files

These include the C++ "unit test" file as well as the expected GiD output files.

avdg81 committed:
Renamed the test directory

avdg81 committed:
Renamed the `.mdpa` file as well as the output and mesh files

markelov208 commented on applications/GeoMechanicsApplication/tests/test_GeoMechanicsApplication.py:
```diff
@@ -110,7 +111,8 @@ def AssembleTestSuites():
                         KratosGeoMechanicsMohrCoulombWithTensionTests,
-                        KratosGeoMechanicsApplyInitialUniformStressFieldTests
+                        KratosGeoMechanicsApplyInitialUniformStressFieldTests,
+                        KratosGeoMechanicsHydraulicHeads,
```
Finally, I know that the trailing coma is recommended.  Before it was difficult to move the last line above. ;)

markelov208 commented on applications/GeoMechanicsApplication/tests/test_nodal_hydraulic_head.py:
```diff
@@ -0,0 +1,73 @@
+import os
+
+
+class KratosGeoMechanicsHydraulicHeads(KratosUnittest.TestCase):
```
AI suggests to add

def setUp(self):
self.test_path = test_helper.get_file_path("test_nodal_hydraulic_head")

avdg81 replied:
Does the AI also provide a rationale for this suggestion, because I'm curious about it? Anyway, I have followed the suggestion, since it removes a local variable in favor of an attribute that could be considered a test case property.

markelov208 replied:
It says to avoid repetition. Using setUp does it once in the background.

markelov208 commented on applications/GeoMechanicsApplication/tests/test_nodal_hydraulic_head.py:
```diff
@@ -0,0 +1,73 @@
+
+
+class KratosGeoMechanicsHydraulicHeads(KratosUnittest.TestCase):
+    def calculate_head(self, head_bottom, head_top, y):
```
AI suggests to add
@staticmethod

avdg81 replied:
Yes, we can indeed make this a `@staticmethod`, since it doesn't need the state of an instance. Done.

markelov208 commented on applications/GeoMechanicsApplication/tests/test_nodal_hydraulic_head.py:
```diff
@@ -0,0 +1,73 @@
+        head_bottom = 1.0
+        head_top = 0.0
+        self.run_and_check_nodal_hydraulic_heads(test_no, head_bottom, head_top)
+
```
Would you like to consider the following

def test_hydraulic_heads(self):
    """Test several hydraulic head scenarios."""
    test_cases = [
        (1, 1.0, 1.0),
        (2, 0.0, 1.0),
        (3, -1.0, 1.0),
        (4, 1.0, 0.0),
    ]
    for test_no, head_bottom, head_top in test_cases:
        with self.subTest(test_no=test_no):
            self.run_and_check_nodal_hydraulic_heads(test_no, head_bottom, head_top)

avdg81 replied:
Excellent suggestion. Basically, what you are saying is to parameterize the test. And so I did, using the `parameterized` module.

markelov208 reviewed (commented):
Hi Anne, a very useful change. AI has suggested a few modifications. Hope some of then are useful.

avdg81 committed:
Merge remote-tracking branch 'origin/master' into geo/improved-head-extrapolation-tests

avdg81 committed:
Processed Gennady's review suggestions

- Use module `parameterized` to simplify the test script.
- Store the test path in an attribute of the test fixture and set it in method `setUp`.
- Made a method a `@staticmethod`.

markelov208 reviewed (dismissed):
Hi Anne, thank you for the changes. I think the PR is good to be merged.

rfaasse commented on applications/GeoMechanicsApplication/tests/test_nodal_hydraulic_head.py:
```diff
@@ -0,0 +1,65 @@
+
+        self.assertVectorAlmostEqual(
+            numerical_hydraulic_heads, analytical_hydraulic_heads, places=5
+        )
```
I think, theoretically, this test would pass if the numberical_hydraulic_heads and analytical_hydraulic_heads are both empty (which might happen if somehow reading the results and reading the coordinates goes wrong. Is it an idea to check that they're not empty?

avdg81 replied:
Good catch. I have added a check on the size of `nodal_coordinates`, which must match the number of nodes in the `.mdpa` file (251).

rfaasse reviewed (commented):
Very clear PR and nice we now have the possibility to run geoflow from python! I only have a single suggestion.

avdg81 committed:
Processed Richard's review suggestion

Make sure that the hydraulic head at each and every node is being checked.

rfaasse reviewed (approved):
Looks good to go to me!
